Add TOSHIBA TZ10xx MCU and Cerevo BlueNinja board support.

f-okuhara opened the pull request:
We added the target "[TOSHIBA TZ1000 Series](http://toshiba.semicon-storage.com/ap-en/product/assp/applite/tz1000.html)" and added the board "Cerevo BlueNinja".
Please merge.

Best regards.

f-okuhara commented:
BlueNinja's interface has been reset pin control through FET.

BlueNinja can be power supply without going through the interface.
Also to be able negate a reset on the If LPC11U35 of power-off.

Reference: [BlueNinja's interface boad's schematic](https://blueninja.cerevo.com/wp-content/themes/blueninja/docs/BlueNinja_CDP-TZ01B_BreakoutBoardSchematic_1503-TZ1-SB-C_20150811.PDF)

c1728p9 commented on source/hic_hal/nxp/lpc11u35/DAP_config.h:
```diff
@@ -181,9 +187,21 @@ Disables the DAP Hardware I/O pins which configures:
+
+#if defined(BLUENINJA_SB)
+    //Release PowerHoldSw
+    LPC_GPIO->CLR[PIN_PWH_PORT] = PIN_PWH;
```
What does PIN_PWH do?  Does it control power to the target?

f-okuhara replied:
Yes.
PIN_PWH is controlled power supply to target.

c1728p9 commented on source/hic_hal/nxp/lpc11u35/gpio.c:
```diff
@@ -82,6 +82,7 @@ void gpio_init(void)
     // enable clock for GPIO port 0
     LPC_SYSCON->SYSAHBCLKCTRL |= (1UL << 6);
     // configure GPIO-LED as output
+#if !defined(BLUENINJA_SB)
```
This is common code so it shouldn't contain board specific logic.  There will need to be feature switch

f-okuhara replied:
I was aware.

c1728p9 commented on test/msd_test.py:
```diff
@@ -296,6 +296,11 @@ def test_mass_storage(workspace, parent_test):
 
+    if target.name == "BlueNinja":
+        test_info.info("Skip the remaining MSD test.")
+        test_info.info("Writing incomplete binary to TZ10xx MCU, CoreSight becomes unresponsive.")
```
Is there any documentation for why the TZ10xx becomes unresponsive when a blank binary is written?  Does this set lock settings or something like that?

toyowata replied:
@c1728p9 I will email you about the TZ10xx docs link.

f-okuhara replied:
MCU manufacturer told me that it is not specified in the document.

The they say:

> When the vector table to write incomplete program is as follows, it is impossible to negotiate CoreSight and debug adapter.
> 
> 1. HardwareFault Exception has occurred
> 2. unrecoverable Exception has occurred
> 3. SystemReset
> 4. Go back to 1
> 
> Restoration is Flash chip erase started with a "SPI Flash erase mode" the MCU.

BlueNinja there is a way to start the hardware to "SPI Flash erase mode".
But, can not be controlled from the interface.

f-okuhara replied:
Providing a Windows 7 environment to identify the critical test items ran the test script.
(In Windows 10 Pro execution of the test script was unstable ...)

Was "Load blank binary + vector table" only is there is no response of CoreSight result.
Other test was able to run without problems.

It fixes to only skip the "Load blank binary + vector table".

f-okuhara commented:
I'm accepted the contributers agreement.
My mbed account is "[okuhara](https://developer.mbed.org/users/okuhara/)".

Best regards.

f-okuhara commented:
I was fix:
- Removed the board switch macro, and added the functional switch macro.
- Added the 'BlueNinja' to TARGET_WITH_BAD_VECTOR_TABLE_LIST.

Still have a probrems to merge?

c1728p9 commented on source/hic_hal/nxp/lpc11u35/DAP_config.h:
```diff
@@ -181,9 +187,21 @@ Disables the DAP Hardware I/O pins which configures:
+#if defined(TARGET_POWER_HOLD)
+    //Release Target PowerHold
+    LPC_GPIO->CLR[PIN_PWH_PORT] = PIN_PWH;
+#endif
```
PORT_OFF gets called when DAPLink is done programming or when a CMSIS-DAP session ends. If this shuts off power to the target you probably don't want that here.

f-okuhara replied:
I want to hold the power only during a programming.
So, it is expected.

I would like you to tell me if there is a better way.

c1728p9 replied:
I was confused about the purpose of PowerHold. I was thinking that clearing power hold turned off power to the MCU when really it just allows the MCU to turn off right? If so then your code is correct and you should ignore my first comment.

f-okuhara replied:
Sorry for insufficient explanation.

<img src="https://docs.google.com/drawings/d/150pRk4Tabfq3527q06CBayEHYfAhrgQ_pG7auyApkhs/pub?w=953&amp;h=642">

Power hold of BlueNinja module the following two elements are responsible.

1. PowerSW (or PowerSW pin of Interface Board)
2. Power management by the software of TZ1000

If PowerSW is OFF, and the hold by the software is released, DC/DC converter must be the Disable.

Power hold by LPC11U35 of the interface is an auxiliary role to power ON only for the duration of the programming.

c1728p9 replied:
Thanks for the info. Now I understand.

c1728p9 commented on source/hic_hal/nxp/lpc11u35/gpio.c:
```diff
@@ -81,7 +81,25 @@ void gpio_init(void)
+    // DAP/MSD/Serial led (green)
+    PIN_DAP_LED_IOCON = PIN_DAP_LED_IOCON_INIT;
+    LPC_GPIO->SET[PIN_DAP_LED_PORT] = PIN_DAP_LED;
+    LPC_GPIO->DIR[PIN_DAP_LED_PORT] |= PIN_DAP_LED;
```
This is the same as "DAP led (green)" below so you should be able to remove "#if defined(INTEGRATED_STATUS_LED)". This will cause PIO0_16 (used for DAP LED, MSD LED, CDC LED on the board) to be initialized to the same value multiple times, but that shouldn't be a problem.

c1728p9 commented:
This is almost ready to go. I added a few more comments, but they are small. Can you attach your test results to the PR? Also, pyOCD support will be needed first - https://github.com/mbedmicro/pyOCD/pull/244

c1728p9 commented on test/hid_test.py:
```diff
@@ -114,6 +114,10 @@ def test_hid(workspace, parent_test):
+        elif target_type == "tz10xx":
+            addr = 0x20000001
+            size = 0x502
+            addr_flash = 0x00000000
```
This is no longer needed if you rebase this PR onto master

c1728p9 commented on test/info.py:
```diff
@@ -105,6 +106,7 @@
     #lpc11u35_arm_watch_efm32_if
     #lpc11u35_arm_watch_nrf51_if
     'Seeed-Arch-Link': 0x9013,
+    'BlueNinja': 0x7010,
```
This will need to be updated to match the format here:
https://github.com/mbedmicro/DAPLink/blob/master/test/info.py#L59

Something like:
(   0x7010,     'lpc11u35_archpro_if',              None,               'BlueNinja'            ),

f-okuhara commented:
I attached test result of this PR.

Run the test.
```
python run_test.py --targetdir C:\Work\Cerevo\TZ1\DAPLink --firmware lpc11u35_blueninja_if --noloadif --force
```

Test target program is BlueNinja.bin and BlueNinja.hex in https://github.com/cerevo/blueninja_DAPLink_test_target .  
Target program was self-made because TOSHIBA TZ1000 series of SDK has not been published.

## requirements.txt
```
cmsis-svd==0.4
colorama==0.3.7
coverage==4.2
elapsedtimer==0.2
enum34==1.1.6
future==0.15.2
intelhex==2.1
Jinja2==2.8
lockfile==0.12.2
MarkupSafe==0.23
mbed-ls==1.2.1
prettytable==0.7.2
project-generator==0.9.9
project-generator-definitions==0.2.29
py==1.4.31
-e git+https://github.com/cerevo/pyOCD.git@dfa4a09bf1e0c2624efd262ccafad9d8263a04c3#egg=pyOCD
pyserial==3.1.1
pytest==3.0.0
pytest-catchlog==1.2.2
pywinusb==0.4.1
PyYAML==3.11
requests==2.11.1
six==1.10.0
websocket-client==0.37.0
xmltodict==0.10.2
```

## summary.txt
```
One or more tests have failed

Git info for test:
  Git SHA: 5bfb6efe1ac0acab732d31de189317f1cc720b28
  Local changes: False

Test settings:
  Load application before test: False
  Load bootloader before test: False
  Run DAPLink specific tests: False
  Run endpoint tests: True

Tested configurations:
  APP=lpc11u35_blueninja_if BL=<None> Board=BlueNinja Target=BlueNinja: Fail

All firmware in package tested
```

## lpc11u35_blueninja_if BlueNinja.txt
```
Test configuration: APP=lpc11u35_blueninja_if BL=<None> Board=BlueNinja Target=BlueNinja
Board: Name=BlueNinja Unique ID=7010000008b02a5300000000000000000000000097969902
Application: Name=lpc11u35_blueninja_if
Bootloader: None
Target: Name=BlueNinja

Test: lpc11u35_blueninja_if BlueNinja: Failure
  Info: Board: Name=BlueNinja Unique ID=7010000008b02a5300000000000000000000000097969902
  Info: Application: Name=lpc11u35_blueninja_if
  Info: Bootloader: None
  Info: Target: Name=BlueNinja
  Test: test_endpoints: Failure
    Test: HID test: Pass
      Info: 



----- FLASH NEW BINARY -----
      Info: 



------ GET Unique ID ------
      Info: Unique ID: 7010000008b02a5300000000000000000000000097969902
      Info: 



------ TEST READ / WRITE CORE REGISTER ------
      Info: initial pc: 0x8
      Info: now pc: 0x3D82
      Info: initial pc value rewritten: 0x8
      Info: MSP = 0x20000100; PSP = 0x00000000
      Info: CONTROL = 0x00; FAULTMASK = 0x00; BASEPRI = 0x00; PRIMASK = 0x00
      Info: New PRIMASK = 0x01
      Info: Restored PRIMASK = 0x00
      Info: 



------ TEST HALT / RESUME ------
      Info: resume
      Info: halt
      Info: HALT: pc: 0xA
      Info: 



------ TEST STEP ------
      Info: reset and halt
      Info: HALT: pc: 0x8
      Info: step
      Info: STEP: pc: 0xA
      Info: step
      Info: STEP: pc: 0xC
      Info: step
      Info: STEP: pc: 0xE
      Info: step
      Info: STEP: pc: 0x10
      Info: 



------ TEST READ / WRITE MEMORY ------
      Info: READ32/WRITE32
      Info: write32 0xF6E6118B at 0x10000001
      Info: read32 at 0x10000001: 0xF6E6118B
      Info: 

READ16/WRITE16
      Info: write16 0x39E5 at 0x10000003
      Info: read16 at 0x10000003: 0x39E5
      Info: 

READ8/WRITE8
      Info: write8 0xFE at 0x10000002
      Info: read8 at 0x10000002: 0xFE
      Info: 



------ TEST READ / WRITE MEMORY BLOCK ------
      Info: TEST PASSED
      Info: 



------ TEST RESET ------
      Info: pc: 0xC
      Info: pc: 0xE
      Info: pc: 0x10
      Info: pc: 0x12
      Info: pc: 0x14
      Info: 



------ TEST PROGRAM/ERASE PAGE ------
      Info: TEST PASSED
      Info: 



----- Restoring image -----
      Info: HID test complete
    Test: Serial test: Failure
      Info: Testing serial port COM54
      Failure: Block test failed
      Info: Testing baud 9600
      Info: Pass
      Info: Testing baud 14400
      Info: Pass
      Info: Testing baud 19200
      Info: Pass
      Info: Testing baud 28800
      Info: Pass
      Info: Testing baud 38400
      Info: Pass
      Info: Testing baud 56000
      Failure: Fail on baud 56000
      Info: Testing baud 57600
      Info: Pass
      Info: Testing baud 115200
      Info: Pass
      Info: Timing test baud 28800
      Info: Pass
      Info: Timing test baud 38400
      Info: Pass
      Info: Timing test baud 56000
      Info: Pass
      Info: Timing test baud 57600
      Info: Pass
      Info: Timing test baud 115200
      Info: Pass
    Test: test_mass_storage: Pass
      Test: Shutil hex file load: Pass
        Info: Loading took 7.9240000248s
        Info: Programming rate 9118.62692755B/s
        Test: wait_for_remount: Pass
          Info: unmount took 1.18600010872 s
          Info: mount took 2.16799998283 s
        Test: test_fs: Pass
          Info: chkdsk returned 0
        Test: test_fs_contents: Pass
          Info: File D:\MBED.HTM valid
          Info: File D:\DETAILS.TXT valid
          Test: test_details_txt: Pass
            Test: parse_kvp_file: Pass
        Test: test_details_txt: Pass
          Test: parse_kvp_file: Pass
        Test: test_fs: Pass
          Info: chkdsk returned 0
        Info: Data matches
      Test: Load hex with flushes: Pass
        Info: Loading took 14.3840000629s
        Info: Programming rate 5023.35926612B/s
        Info: MSD transfer rate 14133.5511061B/s
        Test: wait_for_remount: Pass
          Info: unmount took 0.950999975204 s
          Info: mount took 2.29299998283 s
        Test: test_fs: Pass
          Info: chkdsk returned 0
        Test: test_fs_contents: Pass
          Info: File D:\MBED.HTM valid
          Info: File D:\DETAILS.TXT valid
          Test: test_details_txt: Pass
            Test: parse_kvp_file: Pass
        Test: test_details_txt: Pass
          Test: parse_kvp_file: Pass
        Test: test_fs: Pass
          Info: chkdsk returned 0
        Info: Data matches
      Test: Load blank binary: Pass
        Info: Loading took 0.0320000648499s
        Info: MSD transfer rate 255999.481202B/s
        Test: wait_for_remount: Pass
          Info: unmount took 22.0269999504 s
          Info: mount took 2.16899991035 s
        Test: test_fs: Pass
          Info: chkdsk returned 0
        Test: test_fs_contents: Pass
          Info: File D:\MBED.HTM valid
          Info: File D:\DETAILS.TXT valid
          Info: File D:\FAIL.TXT valid
          Test: test_details_txt: Pass
            Test: parse_kvp_file: Pass
        Test: test_details_txt: Pass
          Test: parse_kvp_file: Pass
        Test: test_fs: Pass
          Info: chkdsk returned 0
        Info: Failure as expected: "The transfer timed out."
      Test: Extra Files: Pass
        Info: Loading took 7.87800002098s
        Info: Programming rate 9171.87100883B/s
        Info: MSD transfer rate 25805.6612666B/s
        Test: wait_for_remount: Pass
          Info: unmount took 1.68499994278 s
          Info: mount took 1.66900014877 s
        Test: test_fs: Pass
          Info: chkdsk returned 0
        Test: test_fs_contents: Pass
          Info: File D:\MBED.HTM valid
          Info: File D:\DETAILS.TXT valid
          Test: test_details_txt: Pass
            Test: parse_kvp_file: Pass
        Test: test_details_txt: Pass
          Test: parse_kvp_file: Pass
        Test: test_fs: Pass
          Info: chkdsk returned 0
        Info: Data matches
      Test: Load good file to restore state: Pass
        Info: Loading took 7.87800002098s
        Info: Programming rate 9171.87100883B/s
        Info: MSD transfer rate 25805.6612666B/s
        Test: wait_for_remount: Pass
          Info: unmount took 1.17000007629 s
          Info: mount took 2.27799987793 s
        Test: test_fs: Pass
          Info: chkdsk returned 0
        Test: test_fs_contents: Pass
          Info: File D:\MBED.HTM valid
          Info: File D:\DETAILS.TXT valid
          Test: test_details_txt: Pass
            Test: parse_kvp_file: Pass
        Test: test_details_txt: Pass
          Test: parse_kvp_file: Pass
        Test: test_fs: Pass
          Info: chkdsk returned 0
        Info: Data matches
```

Some of the serial test will fail...

c1728p9 commented:
Hi @f-okuhara, if you update the target to support 56000 baud it should be able to fix one of the failures. You might need a logic analyzer to determine the cause of the first failure. If I can find time I could also take a look myself.

c1728p9 commented on test/hid_test.py:
```diff
@@ -15,7 +15,7 @@
 # limitations under the License.
 #
- 
+
```
nit - unnecessary witespace

c1728p9 commented on test/info.py:
```diff
@@ -15,7 +15,7 @@
 # limitations under the License.
 #
- 
+
```
nit - unnecessary witespace

c1728p9 commented on test/info.py:
```diff
@@ -56,6 +56,7 @@
     (   0x5050,     'lpc11u35_arm_watch_stm32f411_if',  None,               None                            ),
     (   0x5051,     'lpc11u35_arm_watch_efm32_if',      None,               None                            ),
     (   0x5052,     'lpc11u35_arm_watch_nrf51_if',      None,               None                            ),
+	(   0x7010,     'lpc11u35_blueninja_if',            None,               'BlueNinja'                     ),
```
The target 'BlueNinja' can't be built with the compile API since it is not in mbed-os yet. You'll need to replace 'BlueNinja' with None until support is added. Just leave a comment to the side indicating which target should be used when ready.

c1728p9 commented:
Thanks for the update @f-okuhara, I'll start testing.

c1728p9 commented:
@mbed-bot: TEST

c1728p9 commented:
@mbed-bot: TEST

c1728p9 commented:
Test failed with this error:
"The file records/board/blueninja_sb.yaml referenced in main yaml doesn't exist."

c1728p9 commented:
@mbed-bot: TEST
